source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -130,17 +142,27 @@ void AggressiveDCEPass::ComputeBlock2HeaderMaps(
   currentBranchInst.push(nullptr);
   currentMergeBlockId.push(0);
   for (auto bi = structuredOrder.begin(); bi != structuredOrder.end(); ++bi) {
+    // If leaving an if or loop, update stacks
     if ((*bi)->id() == currentMergeBlockId.top()) {
       currentMergeBlockId.pop();
       currentMergeInst.pop();
       currentBranchInst.pop();
     }
-    block2headerMerge_[*bi] = currentMergeInst.top();
-    block2headerBranch_[*bi] = currentBranchInst.top();
     ir::Instruction* mergeInst;
     ir::Instruction* branchInst;
     uint32_t mergeBlockId;
-    if (IsStructuredIfHeader(*bi, &mergeInst, &branchInst, &mergeBlockId)) {
+    // If there is live code in loop header, the loop is live
+    if (IsStructuredHeader(*bi, SpvOpLoopMerge, &mergeInst, &branchInst,
```

Comment:
I would consider reverting IsStructuredHeader to return true for either case and then conditionally block2HeaderMerge and block2HeaderBranch instead of invoking IsStructuredHeader twice.

Reply:
Got ride of IsStructuredHeader completely and now call IsStructuredIfOrLoopHeader once.